File: tests/houses_token_keeper/claim_reward.ts
import * as anchor from '@coral-xyz/anchor';
import { Program } from '@coral-xyz/anchor';
import { HousesTokenKeeper } from '../../target/types/houses_token_keeper';
import { getKeypairFromFile } from '../../utils/getKeypairFromFile';
import { requestAirdrop } from '../../helpers/requestAirdrop';
import { PublicKey } from '@solana/web3.js';
import {
  TOKEN_PROGRAM_ID,
  ASSOCIATED_TOKEN_PROGRAM_ID,
  mintTo,
} from '@solana/spl-token';
import { getLocalMint } from '../../helpers/getLocalMint';
import { getTokenKeeperAccounts } from '../../helpers/houses_token_keeper/getTokenKeeperAccounts';
import { getStakeAccounts } from '../../helpers/houses_stake/getStakeAccounts';
import { HousesStake } from '../../target/types/houses_stake';
import { getRandomEmail } from '../../utils/getRandomEmail';
import { assert } from 'chai';
import { getTokenAmount } from '../../helpers/getTokenAmount';
import { callAddReward } from '../houses_stake/add_reward';
import { cpiStake } from './stake';

export const cpiClaimReward = async (email) => {
  const program = anchor.workspace
    .HousesTokenKeeper as Program<HousesTokenKeeper>;
  const { userPda, userRewardTokenAccount } = await getTokenKeeperAccounts(
    email
  );
  const [stakePda] = PublicKey.findProgramAddressSync(
    [Buffer.from('stake', 'utf8'), userPda.toBuffer()],
    anchor.workspace.HousesStake.programId
  );
  const adminAccount = getKeypairFromFile(
    '/tests/testAccountsLocal/payer.json'
  );

  const {
    dataPda,
    rewardTokenAccount,
    mint: tokenMint,
    rewardMint,
  } = await getStakeAccounts(adminAccount);

  await program.methods
    .claimReward(email)
    .accounts({
      stakeProgram: anchor.workspace.HousesStake.programId,
      tokenProgram: TOKEN_PROGRAM_ID,
      systemProgram: anchor.web3.SystemProgram.programId,
      associatedTokenProgram: ASSOCIATED_TOKEN_PROGRAM_ID,
      rewardMint,
      tokenMint,
      stakePda,
      dataPda,
      userPda,
      rewardTokenAccount,
      userRewardTokenAccount: userRewardTokenAccount.address,
      authority: adminAccount.publicKey,
    })
    .signers([adminAccount])
    .rpc();
};

describe('cpi claim reward', () => {
  const program = anchor.workspace
    .HousesTokenKeeper as Program<HousesTokenKeeper>;

  const stakeProgram = anchor.workspace.HousesStake as Program<HousesStake>;

  anchor.setProvider(anchor.AnchorProvider.env());
  const adminAccount = getKeypairFromFile(
    '/tests/testAccountsLocal/payer.json'
  );
  const connection = program.provider.connection;
  const tokenMint = getLocalMint();
  const mintAuthority = getKeypairFromFile(
    '/tests/testAccountsLocal/mintAuthority.json'
  );
  const email = getRandomEmail();

  before(async () => {
    await requestAirdrop(connection, adminAccount.publicKey, 5 * 10e5);
    const { userPdaTokenAccount } = await getTokenKeeperAccounts(email);
    await mintTo(
      connection,
      adminAccount,
      tokenMint,
      userPdaTokenAccount.address,
      mintAuthority,
      1000
    );
    await callAddReward(1);
    await cpiStake(email, 500);
    await callAddReward(1000);
  });

  it('claim reward', async () => {
    const { userPda, userRewardTokenAccount } = await getTokenKeeperAccounts(
      email
    );
    const [stakePda] = PublicKey.findProgramAddressSync(
      [Buffer.from('stake', 'utf8'), userPda.toBuffer()],
      anchor.workspace.HousesStake.programId
    );

    let { stacked } = await stakeProgram.account.stakePda.fetch(stakePda);
    assert.equal(stacked.toNumber(), 500);
    assert.equal(await getTokenAmount(userRewardTokenAccount), BigInt(0));
    await cpiClaimReward(email);
    assert.equal(await getTokenAmount(userRewardTokenAccount), BigInt(1000));

    ({ stacked } = await stakeProgram.account.stakePda.fetch(stakePda));
    assert.equal(stacked.toNumber(), 0);
  });
});

File: programs/houses_token_keeper/src/methods/claim_reward.rs
use anchor_lang::prelude::*;
use anchor_spl::{token::{Mint, Token, TokenAccount}, associated_token::AssociatedToken};
use houses_stake::cpi::accounts::ClaimReward as CpiClaimReward;
use houses_stake::types::{StakePda, Data};
use houses_stake::program::HousesStake;

use crate::types::*;

pub fn claim_reward(ctx: Context<ClaimReward>, email: String) -> Result<()> {
    let stake_program = ctx.accounts.stake_program.to_account_info();
    let seeds = &[email.as_ref(), ctx.accounts.authority.key.as_ref(), &[*ctx.bumps.get("user_pda").unwrap()]];
    let signer = &[&seeds[..]];
    let cpi_accounts = CpiClaimReward {
        token_program: ctx.accounts.token_program.to_account_info(),
        system_program: ctx.accounts.system_program.to_account_info(),
        associated_token_program: ctx.accounts.associated_token_program.to_account_info(),
        token_mint: ctx.accounts.token_mint.to_account_info(),
        reward_mint: ctx.accounts.reward_mint.to_account_info(),
        stake_pda: ctx.accounts.stake_pda.to_account_info(),
        data_pda: ctx.accounts.data_pda.to_account_info(),
        reward_token_account: ctx.accounts.reward_token_account.to_account_info(),
        user_reward_token_account: ctx.accounts.user_reward_token_account.to_account_info(),
        user: ctx.accounts.user_pda.to_account_info(),
    };
    let cpi_ctx = CpiContext::new_with_signer(stake_program, cpi_accounts, signer);
    houses_stake::cpi::claim_reward(cpi_ctx)?;
    Ok(())
    }

#[derive(Accounts)]
#[instruction(email: String)]
pub struct ClaimReward<'info> {
    pub token_program: Program<'info, Token>,
    pub system_program: Program<'info, System>,
    pub associated_token_program: Program<'info, AssociatedToken>,
    #[account(
        address = STAKE_PROGRAM_ADDRESS.parse::<Pubkey>().unwrap(),
    )]
    pub stake_program: Program<'info, HousesStake>,

    #[account(
        address = TOKEN_MINT_ADDRESS.parse::<Pubkey>().unwrap(),
    )]
    pub token_mint: Account<'info, Mint>,

    #[account(
        address = REWARD_MINT_ADDRESS.parse::<Pubkey>().unwrap(),
    )]
    pub reward_mint: Account<'info, Mint>,

    #[account(mut)]
    pub stake_pda: Account<'info, StakePda>,

    #[account(mut)]
    pub data_pda: Account<'info, Data>,

    #[account(
        mut,
        associated_token::mint = reward_mint,
        associated_token::authority = data_pda
    )]
    pub reward_token_account: Account<'info, TokenAccount>,

    #[account(
        init_if_needed,
        payer = authority,
        associated_token::mint = reward_mint,
        associated_token::authority = user_pda)]
    pub user_reward_token_account: Account<'info, TokenAccount>,

    #[account(init_if_needed,
        space = 8 + 1,
        payer = authority,
        // constraint rethink
        seeds = [ email.as_ref(), authority.key.as_ref() ],
        bump)]
    pub user_pda: Account<'info, UserPda>,

    #[account(mut,
        address = AUTHORITY_ADDRESS.parse::<Pubkey>().unwrap(),
    )]
    pub authority: Signer<'info>,
}
